File: app/src/main/java/com/cyl/musiclake/ui/music/discover/DiscoverPresenter.java
package com.cyl.musiclake.ui.music.discover;

import android.content.Context;

import com.cyl.musicapi.BaseApiImpl;
import com.cyl.musicapi.bean.ListItem;
import com.cyl.musiclake.api.MusicUtils;
import com.cyl.musiclake.api.baidu.BaiduApiServiceImpl;
import com.cyl.musiclake.base.BasePresenter;
import com.cyl.musiclake.bean.Music;
import com.cyl.musiclake.bean.Playlist;
import com.cyl.musiclake.net.ApiManager;
import com.cyl.musiclake.net.RequestCallBack;
import com.cyl.musiclake.utils.LogUtil;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

/**
 * Des    :
 * Author : master.
 * Date   : 2018/5/20 .
 */
public class DiscoverPresenter extends BasePresenter<DiscoverContract.View> implements DiscoverContract.Presenter {

    private List<Playlist> neteaseLists = new ArrayList<>();
    private String[] charts = new String[]{"云音乐新歌榜"
            , "云音乐热歌榜"
            , "网易原创歌曲榜"
            , "云音乐飙升榜"
            , "云音乐电音榜"
            , "UK排行榜周榜"
            , "美国Billboard周榜 "
            , "KTV嗨榜 "
            , "iTunes榜 "
            , "Hit FM Top榜 "
            , "日本Oricon周榜 "
            , "韩国Melon排行榜周榜 "
            , "韩国Mnet排行榜周榜 "
            , "韩国Melon原声周榜 "
            , "中国TOP排行榜(港台榜) "
            , "中国TOP排行榜(内地榜)"
            , "香港电台中文歌曲龙虎榜 "
            , "华语金曲榜"
            , "中国嘻哈榜"
            , "法国 NRJ EuroHot 30 周榜"
            , "台湾Hito排行榜 "
            , "Beatport全球电子舞曲榜"};

    @Inject
    public DiscoverPresenter() {
    }

    @Override
    public void loadBaidu() {
        ApiManager.request(BaiduApiServiceImpl.INSTANCE.getOnlinePlaylist(), new RequestCallBack<List<Playlist>>() {
            @Override
            public void success(List<Playlist> result) {
                mView.showBaiduCharts(result);
            }

            @Override
            public void error(String msg) {

            }
        });
    }

    @Override
    public void loadNetease() {
        for (int i = 0; i < charts.length; i++) {
            neteaseLists.add(new Playlist(String.valueOf(i), charts[i]));
            int finalI = i;
            BaseApiImpl.Companion.getInstance(mView.getContext()).getTopList(String.valueOf(i), topList -> {
                Playlist playlist = new Playlist();
                playlist.setId(String.valueOf(finalI));
                playlist.setName(topList.getData().getName());
                playlist.setCount(topList.getData().getPlayCount());
                playlist.setCoverUrl(topList.getData().getCover());
                playlist.setDes(topList.getData().getDescription());
                List<Music> musicList = new ArrayList<>();
                if (topList.getData().getList().size() > 0) {
                    for (ListItem item : topList.getData().getList()) {
                        Music music = MusicUtils.INSTANCE.getMusic(item);
                        musicList.add(music);
                    }
                }
                playlist.setMusicList(musicList);
                neteaseLists.set(finalI, playlist);
                mView.showNeteaseCharts(neteaseLists);
                return null;
            });
        }
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/discover/DiscoverPresenter.kt
package com.cyl.musiclake.ui.music.discover

import com.cyl.musicapi.netease.BannerResult
import com.cyl.musiclake.api.music.baidu.BaiduApiServiceImpl
import com.cyl.musiclake.api.music.netease.NeteaseApiServiceImpl
import com.cyl.musiclake.ui.base.BasePresenter
import com.cyl.musiclake.bean.Artist
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.Playlist
import com.cyl.musiclake.api.net.ApiManager
import com.cyl.musiclake.api.net.RequestCallBack
import javax.inject.Inject

/**
 * Des    :
 * Author : master.
 * Date   : 2018/5/20 .
 */
class DiscoverPresenter @Inject
constructor() : BasePresenter<DiscoverContract.View>(), DiscoverContract.Presenter {

    override fun loadBaidu() {
        ApiManager.request(BaiduApiServiceImpl.getOnlinePlaylist(), object : RequestCallBack<MutableList<Playlist>> {
            override fun success(result: MutableList<Playlist>) {
                mView?.showBaiduCharts(result)
            }

            override fun error(msg: String) {
                mView?.showEmptyView(msg)
                mView?.showBaiduCharts(mutableListOf())
            }
        })
    }

    /**
     * 加载网易排行榜（0歌曲）
     */
    override fun loadNetease(tag: String) {
        val observable = NeteaseApiServiceImpl.getTopPlaylists(tag, 30)
        ApiManager.request(observable, object : RequestCallBack<MutableList<Playlist>> {
            override fun success(result: MutableList<Playlist>) {
                mView?.showNeteaseCharts(result)
            }

            override fun error(msg: String) {
                mView?.showEmptyView(msg)
                mView?.showNeteaseCharts(mutableListOf())
            }
        })
    }

    override fun loadArtists() {
        ApiManager.request(NeteaseApiServiceImpl.getTopArtists(30, 0), object : RequestCallBack<MutableList<Artist>> {
            override fun success(result: MutableList<Artist>) {
                mView?.showArtistCharts(result)
            }

            override fun error(msg: String) {
                mView?.showEmptyView(msg)
                mView?.showArtistCharts(mutableListOf())
            }
        })
    }

    override fun loadRaios() {
        loadBannerView()
        val observable = BaiduApiServiceImpl.getRadioChannel()
        ApiManager.request(observable, object : RequestCallBack<MutableList<Playlist>> {
            override fun success(result: MutableList<Playlist>) {
                mView?.showRadioChannels(result)
            }

            override fun error(msg: String) {
                mView?.showEmptyView(msg)
                mView?.showRadioChannels(mutableListOf())
            }
        })
    }

    fun loadBannerView() {
        val observable = NeteaseApiServiceImpl.getBanners()
        ApiManager.request(observable, object : RequestCallBack<BannerResult> {
            override fun success(result: BannerResult) {
                if (result.code == 200) {
                    mView?.showBannerView(result.banners)
                } else {
                    mView?.showBannerView(mutableListOf())
                }
            }

            override fun error(msg: String) {
                mView?.showEmptyView(msg)
                mView?.showBannerView(mutableListOf())
            }
        })
    }

    /**
     * 获取每日推荐（需登录）
     */
    fun loadRecommendPlaylist() {
        val observable = NeteaseApiServiceImpl.recommendPlaylist()
        ApiManager.request(observable, object : RequestCallBack<MutableList<Playlist>> {
            override fun success(result: MutableList<Playlist>) {
                mView?.showRecommendPlaylist(result)
            }

            override fun error(msg: String) {
                mView?.showEmptyView(msg)
                mView?.showRecommendPlaylist(mutableListOf())
            }
        })
    }

    /**
     * 获取推荐歌单
     */
    fun loadPersonalizedPlaylist() {
        val observable = NeteaseApiServiceImpl.personalizedPlaylist()
        ApiManager.request(observable, object : RequestCallBack<MutableList<Playlist>> {
            override fun success(result: MutableList<Playlist>) {
                mView?.showRecommendPlaylist(result)
            }

            override fun error(msg: String) {
                mView?.showEmptyView(msg)
                mView?.showRecommendPlaylist(mutableListOf())
            }
        })
    }
}
